### DISCUSSION TITLE
clh should check system hugepage account if clh enable hugepage

### DISCUSSION THREAD

**Discussion by liangxianlong** [cloud-hypervisor/cloud-hypervisor#5058] [General]
**Describe the bug**
When I use clh to lanuch a VM with hugepage enabled, the memory size is bigger than hugepage account in system,the VM will lanuch succees.

**To Reproduce**
Set nr_hugepages = 2000, then boot vm with "hugepages=on"
```bash
echo 2000 > /proc/sys/vm/nr_hugepages
cloud-hypervisor \
        --kernel ./hypervisor-fw \
        --disk path=focal-server-cloudimg-amd64.raw \
        --cpus boot=4 \
        --memory "size=4096M,hugepages=on,hugepage_size=2M" \
        --net "tap=,mac=,ip=,mask=" \
        --rng
```
Then clh will lanuch succees.

**Version**
```bash
# root @ host-59 in ~ [10:55:28] C:2
$ cloud-hypervisor --version
cloud-hypervisor v25.0
```
But when I use qemu, the qemu will raise error:
```bash
error messages from qemu log: qemu-system-x86_64: unable to map backing store for guest RAM: Cannot allocate memory
```
```

**Comment by sboeuf** [comment]
>When I use clh to lanuch a VM with hugepage enabled, the memory size is bigger than hugepage account in system,the VM will lanuch succees.

That's because when you boot the VM, the guest isn't consuming all the available RAM, meaning the pages aren't allocated. Try running the Cloud Hypervisor VM with `--memory size=4096M,hugepages=on,hugepage_size=2M,prefault=on`, that should fail as expected. The `prefault` option will allocate the pages for the entire mapping (4GiB).

**Reply by rbradford** [reply]
> > When I use clh to lanuch a VM with hugepage enabled, the memory size is bigger than hugepage account in system,the VM will lanuch succees.
> 
> That's because when you boot the VM, the guest isn't consuming all the available RAM, meaning the pages aren't allocated. Try running the Cloud Hypervisor VM with `--memory size=4096M,hugepages=on,hugepage_size=2M,prefault=on`, that should fail as expected. The `prefault` option will allocate the pages for the entire mapping (4GiB).

I thought `MAP_PREFAULT` also did this but the man page contradicts this:

```
       MAP_POPULATE (since Linux 2.5.46)
              Populate (prefault) page tables for a mapping.  For a file mapping, this causes read-ahead on the file.  This will help to  reduce  blocking
              on  page  faults  later.   The mmap() call doesn't fail if the mapping cannot be populated (for example, due to limitations on the number of
              mapped huge pages when using MAP_HUGETLB).  MAP_POPULATE is supported for private mappings only since Linux 2.6.23.
```

**Reply by sboeuf** [reply]
Oh! Thanks @rbradford for spotting this!
@liangxianlong then the correct way to make it fail is by running some workload in the guest consuming all the guest RAM. You can try with `stress -m 16` I guess.
